controllers/utils/multicloud_util.go

```diff
@@ -0,0 +1,330 @@
+}
+
+func EnsureManagedClusterActionForInstallPlan(
+	ctx context.Context, c client.Client, name string, namespace string,
+	installPlan operatorsv1alpha1.InstallPlan, Log logr.Logger) (*actionv1beta1.ManagedClusterAction, error) {
+
+	mcaForInstallPlan := &actionv1beta1.ManagedClusterAction{}
+	if err := c.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, mcaForInstallPlan); err != nil {
+		// If the specific managedClusterAction was not found, create it.
+		if errors.IsNotFound(err) {
+			Log.Info("[EnsureManagedClusterActionForInstallPlan] MCA doesn't exist, create it",
+				"name", name, "namespace", namespace)
+			actionMeta := metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			}
+			actionSpec, err := NewManagedClusterActionForInstallPlanSpec(installPlan)
+			if err != nil {
+				return nil, err
+			}
+			mcaForInstallPlan = &actionv1beta1.ManagedClusterAction{
+				ObjectMeta: actionMeta,
+				Spec:       actionSpec,
+			}
+
+			if err := c.Create(ctx, mcaForInstallPlan); err != nil {
+				return mcaForInstallPlan, nil
+			}
+		}
+	} else {
+		// If the specific managedClusterView was found, update it.
+		Log.Info("[EnsureManagedClusterActionForInstallPlan] MCA already exists, update it",
+			"name", name, "namespace", namespace)
+		actionSpec, err := NewManagedClusterActionForInstallPlanSpec(installPlan)
+		if err != nil {
+			return nil, err
+		}
+		mcaForInstallPlan.Spec = actionSpec
+
+		if err := c.Update(ctx, mcaForInstallPlan); err != nil {
+			return nil, err
+		}
+	}
+
+	return mcaForInstallPlan, nil
+}
+
+func NewManagedClusterActionForInstallPlanSpec(installPlan operatorsv1alpha1.InstallPlan) (actionv1beta1.ActionSpec, error) {
+	installPlanClusterServiceVersionNames, err := json.Marshal(installPlan.Spec.ClusterServiceVersionNames)
+	if err != nil {
+		fmt.Printf("Error trying to Marshal InstallPlan ClusterServiceVersionNames: %s", err.Error())
+	}
+
+	templateContent := fmt.Sprintf(
+		`{"apiVersion": "operators.coreos.com/v1alpha1","kind": "InstallPlan",
+		  "metadata": {"name": "%s","resourceVersion": "%s"},
+		  "spec": {"approval": "Manual","approved": true, "clusterServiceVersionNames": %s}}`,
+		installPlan.ObjectMeta.Name,
+		installPlan.ObjectMeta.ResourceVersion,
+		string(installPlanClusterServiceVersionNames))
+
+	actionSpec := actionv1beta1.ActionSpec{
+		ActionType: actionv1beta1.UpdateActionType,
+		KubeWork: &actionv1beta1.KubeWorkSpec{
+			Resource:       "installplan",
+			Namespace:      installPlan.ObjectMeta.Namespace,
+			ObjectTemplate: runtime.RawExtension{Raw: []byte(templateContent)},
+		},
+	}
+
+	return actionSpec, nil
+}
+
+func CheckManagedClusterActionDone(
+	ctx context.Context, c client.Client, mcaName string, mcaNamespace string, Log logr.Logger) error {
+	// Check the managedClusterAction has completed successfully.
+	mca := &actionv1beta1.ManagedClusterAction{}
+	err := c.Get(ctx, types.NamespacedName{Name: mcaName, Namespace: mcaNamespace}, mca)
```

Comment:
It does work if you query the action within 60 seconds. It gets deleted after that if it has completed.

Reply:
Not sure how feasible this is but we should ask if the ManagedClusterAction api can be updated to allow us to register for a status update. When the MCA status changes (done, error, etc) it would copy the status to our CR before deleting itself.